Workload Generation for Elastic Search and CouchBase Data services

Madanagopal19 opened the pull request:
Adding workload generation script for Elastic search and couch base data service

Execution logs:
https://jenkins.pwx.dev.purestorage.com/job/PDS/job/pds-branch-build-test/96/
https://jenkins.pwx.dev.purestorage.com/job/PDS/job/pds-branch-build-test/99/console - recent run

aetos link - http://aetos.pwx.purestorage.com/resultSet/testSetID/94425/testCaseID/396130

dbhatnagar-px commented on drivers/pds/lib/utils.go:
```diff
@@ -1487,12 +1551,32 @@ func CreateDataServiceWorkloads(dataServiceName string, deploymentID string, sca
 			return nil, nil, err
 		}
+	case elasticSearch:
+		esCommand := "while true; do esrally race --track=geonames --target-hosts=" + dnsEndpoint + " --pipeline=benchmark-only --test-mode --kill-running-processes --client-options=\"timeout:60,use_ssl:false,verify_certs:false,basic_auth_user:'elastic',basic_auth_password:" + pdsPassword + "\"; done"
```
Let's have the options to this command configurable - for tuning the workload better in future tests like scale, longevity.

Madanagopal19 replied:
Done

dbhatnagar-px commented on tests/pds/dataservice_test.go:
```diff
@@ -337,6 +340,24 @@ var _ = Describe("{ScaleUPDataServices}", func() {
+						pod, dep, err = pdslib.CreateDataServiceWorkloads(ds.Name, deployment.GetId(), "", "", deploymentName, namespace)
+						log.FailOnError(err, "Error while genearating workloads")
+					}
+					if ds.Name == elasticSearch {
```
Is there a reason we are running Elasticsearch twice ?

Madanagopal19 replied:
It was added by mistake and its a good catch, Thanks and I have removed it.

lsrinivas-pure commented on drivers/pds/lib/utils.go:
```diff
@@ -1181,8 +1199,61 @@ func CreateIndependentMySqlApp(ns string, podName string, appImage string, pvcNa
+
+	pod, err := k8sCore.CreatePod(podSpec)
+	if err != nil {
+		log.Errorf("An Error Occured while creating %v", err)
```
just return the err and log it at the calling function

Madanagopal19 replied:
Done

lsrinivas-pure commented on drivers/pds/lib/utils.go:
```diff
@@ -1181,8 +1199,61 @@ func CreateIndependentMySqlApp(ns string, podName string, appImage string, pvcNa
+
+	err = k8sCore.ValidatePod(pod, timeOut, timeInterval)
+	if err != nil {
+		log.Errorf("An Error Occured while validating the pod %v", err)
```
same as above

Madanagopal19 replied:
Done

lsrinivas-pure commented on drivers/pds/lib/utils.go:
```diff
@@ -1479,20 +1550,39 @@ func CreateDataServiceWorkloads(dataServiceName string, deploymentID string, sca
 			return nil, nil, err
 		}
+	case elasticSearch:
+		esCommand := "while true; do esrally race --track=geonames --target-hosts=" + dnsEndpoint + " --pipeline=benchmark-only --test-mode --kill-running-processes --client-options=\"timeout:" + params.TimeOut + ",use_ssl:" + params.UseSSL + ",verify_certs:" + params.VerifyCerts + ",basic_auth_user:" + params.User + ",basic_auth_password:" + pdsPassword + "\"; done"
```
use fmt.Sprintf that will clear and readable

Madanagopal19 replied:
Done

lsrinivas-pure commented on drivers/pds/lib/utils.go:
```diff
@@ -1479,20 +1550,39 @@ func CreateDataServiceWorkloads(dataServiceName string, deploymentID string, sca
 	case cassandra:
-		cassCommand := deploymentName + " write no-warmup n=1000000 cl=ONE -mode user=pds password=" + pdsPassword + " native cql3 -col n=FIXED\\(5\\) size=FIXED\\(64\\)  -pop seq=1..1000000 -node " + dnsEndpoint + " -port native=9042 -rate auto -log file=/tmp/" + deploymentName + ".load.data -schema \"replication(factor=3)\" -errors ignore; cat /tmp/" + deploymentName + ".load.data"
-		dep, err = CreatecassandraWorkload(cassCommand, deploymentName, namespace)
+		cassCommand := params.DeploymentName + " write no-warmup n=1000000 cl=ONE -mode user=pds password=" + pdsPassword + " native cql3 -col n=FIXED\\(5\\) size=FIXED\\(64\\)  -pop seq=1..1000000 -node " + dnsEndpoint + " -port native=9042 -rate auto -log file=/tmp/" + params.DeploymentName + ".load.data -schema \"replication(factor=3)\" -errors ignore; cat /tmp/" + params.DeploymentName + ".load.dat ...
```
use fmt.Sprintf that will clear and readable

Madanagopal19 replied:
Done

lsrinivas-pure commented on tests/pds/dataservice_test.go:
```diff
@@ -767,6 +748,57 @@ var _ = Describe("{DeployAllDataServices}", func() {
 })
 
+func RunWorkloads(params pdslib.WorkloadGenerationParams, ds PDSDataService) {
+	params.DataServiceName = ds.Name
```
see if you can move this pds common

Madanagopal19 replied:
Done

Madanagopal19 commented on tests/pds/dataservice_test.go:
```diff
@@ -370,11 +351,11 @@ var _ = Describe("{ScaleUPDataServices}", func() {
 				Step("Delete the workload generating deployments", func() {
-					if ds.Name == cassandra || ds.Name == postgresql || ds.Name == redis || ds.Name == rabbitmq {
-						if ds.Name == cassandra || ds.Name == postgresql {
+					if ds.Name == cassandra || ds.Name == postgresql || ds.Name == redis || ds.Name == rabbitmq || ds.Name == elasticSearch {
```
Done

Madanagopal19 commented on tests/pds/dataservice_test.go:
```diff
@@ -767,6 +748,57 @@ var _ = Describe("{DeployAllDataServices}", func() {
+
+		log.Infof("Running Workloads on DataService %v ", ds.Name)
+		pod, dep, err = pdslib.CreateDataServiceWorkloads(params)
+		log.FailOnError(err, "Error while genearating workloads")
```
Done

Madanagopal19 commented on tests/pds/dataservice_test.go:
```diff
@@ -244,7 +247,7 @@ var _ = Describe("{RestartPDSagentPod}", func() {
 var _ = Describe("{Enable/DisableNamespace}", func() {
 	JustBeforeEach(func() {
-		StartTorpedoTest("ScaleUPDataServices", "Deploys and Scales Up the dataservices", nil, 0)
+		StartTorpedoTest("Enable/DisableNamespace", "Enables and Disables pds on a namespace multiple times ", nil, 0)
```
Done

nikolaypopov commented on drivers/pds/lib/utils.go:
```diff
@@ -735,8 +749,8 @@ func ValidatePDSDeploymentStatus(deployment *pds.ModelsDeployment, healthStatus
 		if !strings.Contains(status.GetHealth(), healthStatus) {
-			err = fmt.Errorf("status %v", status.GetHealth())
-			return false, err
+			log.Errorf("status: %v", status.GetHealth())
+			return false, nil
```
Why are you logging it as error if you not returning it? Either return it as verbose error, don't log it as error or elaborate why this is an error and why you are not returning it as an error

Madanagopal19 replied:
done

nikolaypopov commented on drivers/pds/lib/utils.go:
```diff
@@ -1441,58 +1506,69 @@ func CreateTpccWorkloads(dataServiceName string, deploymentID string, scalefacto
 		if err != nil {
-			log.Errorf("An Error Occured while creating postgresql workload %v", err)
-			return nil, nil, err
+			return nil, nil, fmt.Errorf("error occured whilecreating postgresql workload %v", err)
```
nit: while creating

nikolaypopov commented on drivers/pds/lib/utils.go:
```diff
@@ -1441,58 +1506,69 @@ func CreateTpccWorkloads(dataServiceName string, deploymentID string, scalefacto
 		if err != nil {
-			log.Errorf("An Error Occured while creating rabbitmq workload %v", err)
-			return nil, nil, err
+			return nil, nil, fmt.Errorf("error occured while creating rabbitmq workload %v ", err)
```
nit: remove extra space

Madanagopal19 replied:
done

nikolaypopov commented on drivers/pds/lib/utils.go:
```diff
@@ -1441,58 +1506,69 @@ func CreateTpccWorkloads(dataServiceName string, deploymentID string, scalefacto
 		if err != nil {
-			log.Errorf("An Error Occured while creating redis workload %v", err)
-			return nil, nil, err
+			return nil, nil, fmt.Errorf("error occured while creating redis workload %v ", err)
```
nit: remove extra space

Madanagopal19 replied:
done

nikolaypopov commented on drivers/pds/lib/utils.go:
```diff
@@ -1441,58 +1506,69 @@ func CreateTpccWorkloads(dataServiceName string, deploymentID string, scalefacto
 	if err != nil {
-		log.Errorf("An Error Occured while getting credentials info %v", err)
-		return nil, nil, err
+		return nil, nil, fmt.Errorf("error occured while getting credentials info %v ", err)
```
nit: remove extra space

Madanagopal19 replied:
done

nikolaypopov commented on drivers/pds/lib/utils.go:
```diff
@@ -1441,58 +1506,69 @@ func CreateTpccWorkloads(dataServiceName string, deploymentID string, scalefacto
 		if err != nil {
-			log.Errorf("An Error Occured while creating cassandra workload %v", err)
-			return nil, nil, err
+			return nil, nil, fmt.Errorf("error occured while creating cassandra workload %v ", err)
```
nit: please find and remove ALL extra space from messages

Madanagopal19 replied:
done

nikolaypopov commented on drivers/pds/lib/utils.go:
```diff
@@ -1715,18 +1790,21 @@ func GetAllSupportedDataServices() map[string]string {
+	err = wait.Poll(maxtimeInterval, timeOut, func() (bool, error) {
+		deployment, err = components.DataServiceDeployment.UpdateDeployment(deploymentID, appConfigID, imageID, nodeCount, resourceTemplateID, nil)
+		if err != nil {
+			log.Errorf("An Error Occured while updating deployment %v", err)
+			return false, nil
```
Don't log the error just return it if you need to, and don't you want to return this error? Why are we returning `nil` here?

Madanagopal19 replied:
Removed the error logs

nikolaypopov commented on drivers/pds/lib/utils.go:
```diff
@@ -1715,18 +1790,21 @@ func GetAllSupportedDataServices() map[string]string {
 	}
 
-	return deployment, nil
+	return deployment, err
```
if we got here, it should be `nil`, otherwise we should have failed and returned above, no?

Madanagopal19 replied:
yes, you are right and Thanks for the catch. I have fixed the same

nikolaypopov commented on tests/pds/dataservice_test.go:
```diff
@@ -41,21 +41,21 @@ var _ = Describe("{DeletePDSPods}", func() {
-					if len(deploymentPods) != 0 {
-						deploymentPods = nil
+					if len(pdsPods) != 0 {
+						pdsPods = nil
 					}
 					podList, err = pdslib.GetPods(pdsNamespace)
 					log.FailOnError(err, "Error while getting pods")
 					log.Infof("PDS System Pods")
 					for _, pod := range podList.Items {
 						log.Infof("%v", pod.Name)
-						deploymentPods = append(deploymentPods, pod)
+						pdsPods = append(pdsPods, pod)
 					}
```
This is confusing, why don't we get podList first `pdslib.GetPods(pdsNamespace)`, and then append them to []pdsPods? there is no point to check `len(pdsPods)` if you haven't even got the list of them yet

Madanagopal19 replied:
`pdsPods` is a global variable and its used by other tests also, I expect always `pdsPods []corev1.Pod` to be empty at the start of the test.

nikolaypopov replied:
You should just do this and define pdsPods afrer you get them and get rid of setting it to `nil`
```
					podList, err := pdslib.GetPods(pdsNamespace)
					log.FailOnError(err, "Error while getting pds pods")
					log.Infof("PDS system pods")
					for _, pod := range podList.Items {
						log.Infof("%v", pod.Name)
						pdsPods = append(pdsPods, pod)
					}
```

Madanagopal19 replied:
Test `DeletePDSPods` runs in a for loop (depends on the data services we are passing as a param ) and each time it will populate `pdsPods`. 
For first iteration it will have set of pods say ["xxx", "yyy"] and it will delete those pods as per the test case and new pods will get created in different names ["aaa","bbb"]. In the next iteration we need to delete the new pods but  the `pdsPods` will have the old pod entries so we need to empty `pdsPods` else `pdsPods` will have entries like  ["xxx", "yyy", "aaa", "bbb"] and during pod deletion step the test will fail because pods ["xxx", "yyy"] will not be available. 
hope I have cleared the confusion here.

Madanagopal19 replied:
I have removed that check

nikolaypopov commented on tests/pds/dataservice_test.go:
```diff
@@ -117,10 +117,13 @@ var _ = Describe("{ValidatePDSHealthInCaseOfFailures}", func() {
 
 				log.Infof("PDS DataService Pods")
 				log.Infof("deployment name %v", *deployment.ClusterResourceName)
+				if len(pdsPods) != 0 {
+					pdsPods = nil
+				}
```
Why are we doing this if we know it is empty? We don't need this

Madanagopal19 replied:
`pdsPods` is a global variable and its used by other tests also, in some cases it will not be empty hence Im making it empty at start of the test

nikolaypopov commented on tests/pds/dataservice_test.go:
```diff
@@ -313,30 +316,9 @@ var _ = Describe("{ScaleUPDataServices}", func() {
-					}
+					var params pdslib.WorkloadGenerationParams
+					pod, dep, err = RunWorkloads(params, ds, deployment, namespace)
+					log.FailOnError(err, "Error while genearating workloads")
```
Can we elaborate what workloads we failed to generate here? Something to give us a hint or more info about what app failed deployment
```
log.FailOnError(err, fmt.Sprintf("Error while genearating workloads [%s]", ds.Name))
```

Madanagopal19 replied:
done

nikolaypopov reviewed (changes requested):
There is room for improvement and some changes we need to make, please reach out if you want to discuss

Madanagopal19 commented on drivers/pds/lib/utils.go:
```diff
@@ -1181,8 +1195,59 @@ func CreateIndependentMySqlApp(ns string, podName string, appImage string, pvcNa
+
+	pod, err := k8sCore.CreatePod(podSpec)
+	if err != nil {
+		return nil, fmt.Errorf("error occured while creating pod %v", err)
```
done

Madanagopal19 commented on drivers/pds/lib/utils.go:
```diff
@@ -1181,8 +1195,59 @@ func CreateIndependentMySqlApp(ns string, podName string, appImage string, pvcNa
+
+	err = k8sCore.ValidatePod(pod, timeOut, timeInterval)
+	if err != nil {
+		return nil, fmt.Errorf("error occured while validating pod %v", err)
```
done

lsrinivas-pure commented on tests/pds/dataservice_test.go:
```diff
@@ -370,11 +353,11 @@ var _ = Describe("{ScaleUPDataServices}", func() {
 				Step("Delete the workload generating deployments", func() {
-					if ds.Name == cassandra || ds.Name == postgresql || ds.Name == redis || ds.Name == rabbitmq {
-						if ds.Name == cassandra || ds.Name == postgresql {
+					if Contains(dataServiceDeploymentWorkloads, ds.Name) || Contains(dataServicePodWorkloads, ds.Name) {
```
why do u need this or condition if you separate if condition for dataServiceDeploymentWorkloads and dataServicePodWorkloads again

Madanagopal19 replied:
yeah you are right, I have updated the same

lsrinivas-pure commented on drivers/pds/lib/utils.go:
```diff
@@ -1441,58 +1511,69 @@ func CreateTpccWorkloads(dataServiceName string, deploymentID string, scalefacto
 		if err != nil {
-			log.Errorf("An Error Occured while creating postgresql workload %v", err)
-			return nil, nil, err
+			return nil, nil, fmt.Errorf("error occured while creating postgresql workload %v", err)
```
can you log like this 
fmt.Errorf("error occured while creating postgresql workload, Err: %v", err)

it will be more clear

Madanagopal19 replied:
done

dbhatnagar-px reviewed (approved):
LGTM

Madanagopal19 commented:
> There is room for improvement and some changes we need to make, please reach out if you want to discuss

Hi @nikolaypopov I have addressed the comments
